packages/cache/ARCHITECTURE.md

```diff
@@ -0,0 +1,56 @@
+# TODO: title
+
+## Assumptions
+
+- The `CacheRegistry` will be generated by the "Codegen" system where the full
+  set of query keys are known ahead of time. In Athena's case that will mean
+  that inidivual query ids like `query-id-123` will map to **exactly** the return type of that query.
+  Similarly, `cache.get(someEntityId)` will map to **exactly** the correct type of the "merged model". For an example of generating these type registries, see <https://tsplay.dev/NrnDlN>. This gives strong types for entries whose types are statically known. There is no support for strongly typing dynamic entries -- they will be typed as `unknown` and users will have to typecast or equivalent.
+
+## Features
+
+// TODO: rewrite after writing `save` api
+
+- The cache is serializable. This means e.g. it can be written to and restored from persistent storage. It also means that values must be serializable one way or another.
+
+## Implementation Notes
+
+- 🔥🐉 When cache entries (key,value pairs) are added to the cache (i.e. via `tx.commit()`) particular care is needed.
+  1. The main cache is a `Map<String, WeakRef<ValueType>>`.
+  2. There are additional caches that are `Map<string,ValueType>` to handle the
+     Least-Recently-Used (LRU) and Time-to-Live (TTL) expiration features.
+  3. This means that keys are always strongly held (more on this in a minute) and
+     values are strongly held until they expire, after which they are weakly
+     held by the cache, but may be strongly held by the user (so e.g. a
```

Comment:
To have this working, we will need to ensure the references for the "values" are implemented & maintained correctly between the weekly held and strongly held cache objects, do we foresee anything that can get buggy that we need to be careful about?

Reply:
Got it, we will keep a note of this with implementation & tests